kimchi/src/circuits/polynomials/foreign_field_add/circuitgates.rs

```diff
@@ -0,0 +1,200 @@
+//~ and `result` is an intermediate value that is unused elsewhere (since the final `result`
+//~ must have had the right amount of moduli subtracted along the way, meaning a multiple of the modulus).
+//~
+//~ You could lay this out as a double-width gate for chained foreign additions and a final row, e.g.
+//~
+//~ | col | `ForeignFieldAdd`       | more `ForeignFieldAdd` | final `ForeignFieldAdd` | final `Zero`      |
+//~ | --- | ----------------------- | ---------------------- | ----------------------- | ----------------- |
+//~ |   0 | `left_input_lo`  (copy) | `result_lo` (copy)     | `min_result_lo` (copy)  | `bound_lo` (copy) |
+//~ |   1 | `left_input_mi`  (copy) | `result_mi` (copy)     | `min_result_mi` (copy)  | `bound_mi` (copy) |
+//~ |   2 | `left_input_hi`  (copy) | `result_hi` (copy)     | `min_result_hi` (copy)  | `bound_hi` (copy) |
+//~ |   3 | `right_input_lo` (copy) |  ...                   |  0              (check) |                   |
+//~ |   4 | `right_input_mi` (copy) |  ...                   |  0              (check) |                   |
+//~ |   5 | `right_input_hi` (copy) |  ...                   |  2^88           (check) |                   |
+//~ |   6 | `field_overflow`        |  ...                   |  1              (check) |                   |
+//~ |   7 | `carry_lo`              |  ...                   | `bound_carry_lo`        |                   |
+//~ |   8 | `carry_mi`              |  ...                   | `bound_carry_mi`        |                   |
+//~ |   9 | `sign`                  |  ...                   |  1              (check) |                   |
+//~ |  10 |                         |                        |                         |                   |
+//~ |  11 |                         |                        |                         |                   |
+//~ |  12 |                         |                        |                         |                   |
+//~ |  13 |                         |                        |                         |                   |
+//~ |  14 |                         |                        |                         |                   |
+//~
+//~ We reuse the foreign field addition gate for the final bound check since this is an addition with a
+//~ specific parameter structure. Checking that the correct right input, overflow, and sign are used shall
+//~ be done by copy constraining these values with a public input value.
+
+/// Implementation of the foreign field addition gate
+/// - Operates on Curr and Next rows.
+pub struct ForeignFieldAdd<F>(PhantomData<F>);
+
+impl<F> Argument<F> for ForeignFieldAdd<F>
+where
+    F: FftField,
+{
+    const ARGUMENT_TYPE: ArgumentType = ArgumentType::Gate(GateType::ForeignFieldAdd);
+    const CONSTRAINTS: u32 = 7;
+
+    fn constraint_checks<T: ExprOps<F>>(env: &ArgumentEnv<F, T>) -> Vec<T> {
+        let field_overflow = env.witness_curr(6);
+        let sign = env.witness_curr(9);
```

Comment:
Oh you're right... this was handled within the different gate but now this should be copy constrained to something in the public inout. Yeah, I agree we could just place it where ovf is, but then you wouldn't be able to check that ovf is 1 either, no?

Reply:
Ok actually, since the constraint is that sign must be one or minus one and ovf should be zero or sign, if we just constrain ovf to be one (located in column <7) and the constraint for ovf * (ovf - s) works, then that implicitly gives us the check of sign = 1 for free. Thus I think it is sufficient if we leave the layout as is.

Reply:
`ovf = 0` is ambiguous for `0 < b < min(a,modulus-b-1)`. We still should switch them.

Reply:
Mmmm I am unsure about this case... Can you explain more? I cannot think of any example where the overflow will be zero in the final check. Meaning, since you are adding the minimum result to 2^{264}, this will always cause overflow, right?

Reply:
Oh, I think you are referring to the sign for all the other FFadds, not just the final one right? I get you now 👍🏻